File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/src/grammarGraph/getParentStackId.ts
import { Color } from './colorize.ts';
import type { Colorize } from './grammarGraphTypes.ts';
import type { GraphPointer } from './graphPointer.ts';

export const getParentStackId = (pointer: GraphPointer, col: Colorize): string => {
  const stackIds: string[] = [];
  let parent: GraphPointer | null = pointer.parent;
  while (parent) {
    stackIds.push(
      `${parent.node.meta.stackId},${parent.node.meta.pathId},${parent.node.meta.stepId}`,
    );
    parent = parent.parent;
  }
  const arrow = col('<-', Color.GRAY);
  return stackIds.map((stackId) => col(stackId, Color.RED)).join(arrow);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T181751Z_00b43b48/20260910T095901Z_ef14bd50/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from typing import TYPE_CHECKING, List

from .colorize import Color

if TYPE_CHECKING:
    from .grammar_graph_types import Colorize
    from .graph_pointer import GraphPointer


def get_parent_stack_id(pointer: "GraphPointer", col: "Colorize") -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stack_id},{meta.path_id},{meta.step_id}")
        parent = parent.parent
    arrow = col("<-", Color.GRAY)
    return arrow.join(col(stack_id, Color.RED) for stack_id in stack_ids)
